### ISSUE TITLE
[feature request] dfdaemon support private registry with HTTPS protocol

### ISSUE THREAD

**Issue by lowzj** [dragonflyoss/Dragonfly#205]
---
name: Feature Request
about: Suggest your excellent idea for this project

---

## Why you need it?

It doesn't work when I pull private container images from an image registry with HTTPS protocol by using dragonfly.

The error outputted by dockerd:

```
denied: requested access to the resource is denied
unauthorized: authentication required
```

## How it could be?

Actually, the `dfdaemon` is just a proxy between dockerd and registry. The `dfdaemon` captures every request send from dockerd to registry, filters out all the image layer downloading requests and use `dfget` to download them.

We should investigate HTTPS proxy implementation and make it possible for `dfdaemon` to get http data from tcp data, analysis them and download image layers correctly.

**Comment by allencloud** [comment]
@chenchaobing @zhouhaibing089 Any ideas on this? 😄

**Comment by xh4n3** [comment]
If dfdaemon works as a registry mirror, it can only proxy images on docker hub. It seems it can support https well. To use private registry, we might change docker engine's code. See https://github.com/moby/moby/issues/18818

---

When dfdaemon works as a http proxy for docker engine, it does have this https issue.

For registries that use OSS as storage backend, we can use https for registry itself, and specify insecure option to the oss driver. Then the registry protocol goes in https, all blob-releated traffic goes in http. In this case, Dragonfly could still capture and manipulate blob traffic, while keep registry and authentication infos in secure https. 

In short and IMHO, to support https private registries, we might need to sacrifice blobs traffic's security for Dragonfly's P2P ability.

**Comment by rpardini** [comment]
Interesting. Please check out https://github.com/rpardini/docker-registry-proxy where I implemented a (admittedly hack-ish) way of proxying ALL requests from the Docker daemon, including HTTPS, and to any registry. It involves setting HTTP(S)_PROXY and injecting a root CA.
Lets talk if interested.
